src/log_impl.rs

```diff
@@ -528,6 +569,75 @@ impl Log for Panic {
     fn flush(&self) {}
 }
 
+impl Log for CustomFileLog {
+    fn enabled(&self, _: &log::Metadata) -> bool {
+        true
+    }
+
+    fn log(&self, record: &log::Record) {
+        fallback_on_error(record, |record| {
+            let is_writable = self.is_writable.lock().unwrap_or_else(|e| e.into_inner());
+
+            if *is_writable {
+                if cfg!(feature = "meta-logging-in-format") {
+                    // Formatting first prevents deadlocks on file-logging,
+                    // when the process of formatting itself is logged.
+                    // note: this is only ever needed if some Debug, Display, or other
+                    // formatting trait itself is logging.
+                    let msg = format!("{}{}", record.args(), self.line_sep);
+
+                    let mut writer = self.file_stream.lock().unwrap_or_else(|e| e.into_inner());
+
+                    write!(writer, "{}", msg)?;
+
+                    writer.flush()?;
+                } else {
+                    let mut writer = self.file_stream.lock().unwrap_or_else(|e| e.into_inner());
+
+                    write!(writer, "{}{}", record.args(), self.line_sep)?;
+
+                    writer.flush()?;
+                }
+
+                //now check if log needs to be rotated
```

Comment:
I would usually choose to rotate before logging to the output file, but that might naive of me.

Any particular reason for rotating after logging rather than before? I'm fine with it either way, just wondering.

Reply:
I was thinking of moving this out into a separate timed thread, so that the actual logging does not have any performance hit. It would also lend itself to a size based monitoring also in that case.
What are your thoughts on this?

Reply:
Sorry about the delay here.

With threading, it could definitely be useful, but if I understand things correctly, file logging in general will have hits like this when the the underlying buffer flushes. If we want to prevent this, I think we'd be better served putting all the logging on a separate thread rather than just the re-opening.

Even if we do refresh in a separate thread, we'll block the main thread the next time it logs to the file anyways. Because of this, and file logging blocking in general, I would be inclined to just do it in place.